packages/mobile/src/identity/contactMapping.ts

```diff
@@ -105,164 +93,65 @@ function* updateRecipientsCache(
-  const phoneHashes = e164Numbers.map((phoneNumber) => getPhoneHash(phoneNumber))
-  const results = await attestationsWrapper.lookupPhoneNumbers(phoneHashes)
-  if (!results) {
-    return null
-  }
-
-  const addresses: Array<string | null> = []
-  for (const hash of phoneHashes) {
-    if (results[hash]) {
-      // TODO(Rossy) Add support for handling multiple addresses per number
-      const addressArray = Object.keys(results[hash])
-      const address = addressArray[addressArray.length - 1]
-      addresses.push(address.toLowerCase())
-    } else {
-      addresses.push(null)
-    }
-  }
-
-  if (!addresses || addresses.length !== e164Numbers.length) {
-    throw new Error('Address lookup length did not match numbers list length')
-  }
-
-  return addresses
-}
-
-const isValidAddress = (address: string) =>
-  typeof address === 'string' && !new BigNumber(address).isZero()
-
-export function* fetchAndStoreAddressMappings(
-  attestationsWrapper: AttestationsWrapper,
-  e164Numbers: string[],
-  incrementSyncProgress = true
-) {
+export function* fetchPhoneAddresses({ e164Number }: FetchPhoneAddressesAction) {
   try {
-    Logger.debug(TAG, `Fetch and store address mapping for ${e164Numbers.length} phone numbers`)
-
-    const addresses: Array<string | null> = yield call(retryAsync, getAddresses, 3, [
-      e164Numbers,
-      attestationsWrapper,
+    Logger.debug(TAG + '@fetchPhoneAddresses', `Fetching addresses for number`)
+    // Clear existing entries for those numbers so our mapping consumers
+    // know new status is pending.
+    yield put(updateE164PhoneNumberAddresses({ [e164Number]: undefined }, {}))
+
+    const contractKit = getContractKit()
+    const attestationsWrapper: AttestationsWrapper = yield call([
+      contractKit.contracts,
+      contractKit.contracts.getAttestations,
     ])
 
-    Logger.debug(TAG, `Retrieved ${addresses.length} addresses`)
+    const addresses: string[] | null = yield call(getAddresses, e164Number, attestationsWrapper)
 
     const e164NumberToAddressUpdates: E164NumberToAddressType = {}
     const addressToE164NumberUpdates: AddressToE164NumberType = {}
 
-    for (let i = 0; i < addresses.length; i++) {
-      const address = addresses[i]
-      const e164Number = e164Numbers[i]
-
-      if (address && isValidAddress(address)) {
-        e164NumberToAddressUpdates[e164Number] = address
-        addressToE164NumberUpdates[address] = e164Number
-      } else {
-        // Save invalid/0 addresses to avoid checking again
-        // null means a contact is unverified, whereas undefined means we haven't checked yet
-        e164NumberToAddressUpdates[e164Number] = null
-      }
+    if (!addresses) {
+      Logger.debug(TAG + '@fetchPhoneAddresses', `No addresses for for number`)
+      // Save invalid/0 addresses to avoid checking again
+      // null means a contact is unverified, whereas undefined means we haven't checked yet
+      e164NumberToAddressUpdates[e164Number] = null
```

Comment:
I feel a bit concerned about using `null` and `undefined` to represent different logical states. Even though, that is probably how people should've been using them, in reality it would be hard to catch some potentially dangerous code like this one in the future:
 ```javascript
if (!e164NumberToAddressUpdates[e164Number]) {
    // do stuff
 }
```
WDYT?

Reply:
Yeah I don't love it either but I liked it slightly better than using some other Sentinel value like `-1`.
Then intention is not to have anyone access `e164NumberToAddressUpdates` directly, they should use the `getAddressFromPhoneNumber` helper. How about I add a comment explaining that?

Reply:
Yeah, sounds good!